src/librustc/infer/mod.rs

```diff
@@ -135,6 +147,39 @@ pub struct InferCtxt<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
 
     // This flag is true while there is an active snapshot.
     in_snapshot: Cell<bool>,
+
+    // A set of constraints that regionck must validate. Each
+    // constraint has the form `T:'a`, meaning "some type `T` must
+    // outlive the lifetime 'a". These constraints derive from
+    // instantiated type parameters. So if you had a struct defined
+    // like
+    //
+    //     struct Foo<T:'static> { ... }
+    //
+    // then in some expression `let x = Foo { ... }` it will
+    // instantiate the type parameter `T` with a fresh type `$0`. At
+    // the same time, it will record a region obligation of
+    // `$0:'static`. This will get checked later by regionck. (We
+    // can't generally check these things right away because we have
+    // to wait until types are resolved.)
+    //
+    // These are stored in a map keyed to the id of the innermost
+    // enclosing fn body / static initializer expression. This is
+    // because the location where the obligation was incurred can be
+    // relevant with respect to which sublifetime assumptions are in
+    // place. The reason that we store under the fn-id, and not
+    // something more fine-grained, is so that it is easier for
+    // regionck to be sure that it has found *all* the region
+    // obligations (otherwise, it's easy to fail to walk to a
+    // particular node-id).
+    //
+    // Before running `resolve_regions_and_report_errors`, the creator
+    // of the inference context is expected to invoke
+    // `process_region_obligations` (defined in `self::region_obligations`)
+    // for each body-id in this map, which will process the
+    // obligations within. This is expected to be done 'late enough'
+    // that all type inference variables have been bound and so forth.
+    region_obligations: RefCell<NodeMap<Vec<RegionObligation<'tcx>>>>,
 }
```

Comment:
I don't think this handle snapshots correctly - i.e. are region obligations that were created in a "child" fulfillcx in a probe ignored? `type_known_to_meet_bound` creates a "child" fulfillment context, and I'm not sure it can't be called from typeck snapshots.

If you think nobody should be registering region obligations in a snapshot, which should eventually be the case with MIR regionck, it's a good idea to add an assertion to prevent things blowing up.

Reply:
Good point. I'll add an assertion to check. If necessary, we could make the map snapshottable.

Reply:
I decided to replace this with a vector and support snapshotting.